util/chunk/codec.go

```diff
@@ -189,3 +189,141 @@ func init() {
+//    2.3 Append srcCol.nullBitMap to destCol.nullBitMap.
+// 3. Go to step 1 when the input byte slice is consumed.
+type Decoder struct {
+	intermChk    *Chunk
+	colTypes     []*types.FieldType
+	remainedRows int
+}
+
+// NewDecoder creates a new Decoder object for decode a Chunk.
+func NewDecoder(chk *Chunk, colTypes []*types.FieldType) *Decoder {
+	return &Decoder{intermChk: chk, colTypes: colTypes, remainedRows: 0}
+}
+
+// Decode decodes a Chunk from Decoder, save the remained unused bytes.
+func (c *Decoder) Decode(target *Chunk) {
+	requiredRows := target.RequiredRows() - target.NumRows()
+	// Floor requiredRows to the next multiple of 8
+	requiredRows = (requiredRows + 7) >> 3 << 3
+	if requiredRows > c.remainedRows {
+		requiredRows = c.remainedRows
+	}
+	for i := 0; i < len(c.colTypes); i++ {
+		c.decodeColumn(target, i, requiredRows)
+	}
+	c.remainedRows -= requiredRows
+}
+
+// ResetAndInit resets Decoder, and use data byte slice to init it.
+func (c *Decoder) ResetAndInit(data []byte) {
+	codec := NewCodec(c.colTypes)
+	codec.DecodeToChunk(data, c.intermChk)
+	c.remainedRows = c.intermChk.NumRows()
+}
+
+// IsFinished indicate the data byte slice is consumed.
+func (c *Decoder) IsFinished() bool {
+	return c.remainedRows == 0
+}
+
+func (c *Decoder) decodeColumn(target *Chunk, ordinal int, requiredRows int) {
+	elemLen := getFixedLen(c.colTypes[ordinal])
+	numDataBytes := int64(elemLen * requiredRows)
+	srcCol := c.intermChk.columns[ordinal]
+	destCol := target.columns[ordinal]
+
+	if elemLen == varElemLen {
+		numDataBytes = srcCol.offsets[requiredRows] - srcCol.offsets[0]
+		deltaOffset := destCol.offsets[destCol.length] - srcCol.offsets[0]
+		destCol.offsets = append(destCol.offsets, srcCol.offsets[1:requiredRows+1]...)
+
+		for i := destCol.length + 1; i <= destCol.length+requiredRows; i++ {
+			destCol.offsets[i] = destCol.offsets[i] + deltaOffset
+		}
+		srcCol.offsets = srcCol.offsets[requiredRows:]
+	} else if cap(destCol.elemBuf) < elemLen {
+		destCol.elemBuf = make([]byte, elemLen)
+	}
+
+	numNullBitmapBytes := (requiredRows + 7) >> 3
+	if destCol.length%8 == 0 {
+		destCol.nullBitmap = append(destCol.nullBitmap, srcCol.nullBitmap[:numNullBitmapBytes]...)
+	} else {
+		destCol.appendMultiSameNullBitmap(false, requiredRows)
+		bitMapLen := len(destCol.nullBitmap)
+		// bitOffset indicates the number of elements in destCol.nullBitmap's last byte.
+		bitOffset := destCol.length % 8
+		startIdx := (destCol.length - 1) >> 3
+		for i := 0; i < numNullBitmapBytes; i++ {
+			destCol.nullBitmap[startIdx+i] |= srcCol.nullBitmap[i] << bitOffset
+			// the total number of elements maybe is small than bitMapLen * 8, and last some bits in srcCol.nullBitmap are useless.
+			if bitMapLen > startIdx+i+1 {
+				destCol.nullBitmap[startIdx+i+1] |= srcCol.nullBitmap[i] >> (8 - bitOffset)
+			}
+		}
+	}
+	srcCol.nullBitmap = srcCol.nullBitmap[numNullBitmapBytes:]
```

Comment:
Will there exist some valid bits in `srcCol.nullBitmap[numNullBitmapBytes-1]` ?

Reply:
No. 
If the requiredRows is a multiple of 8, the byte will append to destCol.  
If the requiredRows isn't a multiple of 8,  it indicates the remainedRows will be 0, then reset the intermChk.